mlflow/onnx.py

```diff
@@ -0,0 +1,182 @@
+        # inference engine. Therefore, the conda environment must
+        # include OnnxRuntime
+        "onnxruntime=={}".format(onnxrt.__version__),
+    ],
+    additional_conda_channels=None,
+)
+
+
+def save_model(onnx_model, path, conda_env=None, mlflow_model=Model()):
+    """
+    Save an ONNX model to a path on the local file system.
+
+    :param onnx_model: ONNX model to be saved.
+    :param path: Local path where the model is to be saved.
+    :param conda_env: Either a dictionary representation of a Conda environment or the path to a
+                      Conda environment yaml file. If provided, this decribes the environment
+                      this model should be run in. At minimum, it should specify the dependencies
+                      contained in ``mlflow.onnx.DEFAULT_CONDA_ENV``. If `None`, the default
+                      ``mlflow.onnx.DEFAULT_CONDA_ENV`` environment will be added to the model.
+                      The following is an *example* dictionary representation of a Conda
+                      environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults'],
+                            'dependencies': [
+                                'python=3.6.0',
+                                'onnx=1.4.1',
+                                'onnxmltools=1.4.0',
+                                'onnxruntime=0.3.0'
+                            ]
+                        }
+
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+    """
+    path = os.path.abspath(path)
+    if os.path.exists(path):
+        raise Exception("Path '{}' already exists".format(path))
+    os.makedirs(path)
+    model_data_subpath = "model.onnx"
+    model_data_path = os.path.join(path, model_data_subpath)
+
+    # Save onnx-model
+    onnxmltools.utils.save_model(onnx_model, model_data_path)
+
+    conda_env_subpath = "conda.yaml"
+    if conda_env is None:
+        conda_env = DEFAULT_CONDA_ENV
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    with open(os.path.join(path, conda_env_subpath), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.onnx",
+                        data=model_data_subpath, env=conda_env_subpath)
+    mlflow_model.add_flavor(FLAVOR_NAME, onnx_version=onnx.__version__, data=model_data_subpath)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def _load_model(model_file):
+    onnx_model = onnx.load(model_file)
+    # Check Formation
+    onnx.checker.check_model(onnx_model)
+    return onnx_model
+
+
+class _OnnxModelWrapper:
+    def __init__(self, path):
+        self.onnx_model = _load_model(path)
+        self.rt = onnxrt.InferenceSession(path)
+        assert len(self.rt.get_inputs()) == 1
```

Comment:
@dbczumar It does make sense. However, when I was testing the code with similar setup as the keras tests (iris dataset), I noticed that for this dataset `onnxruntime.get_inputs()` returns only one input (['dense_1_input_01']) which corresponds to the whole dataframe that we pass as input to `predict` in the tests. Thus, for this dataframe, the  `feed_dict = {
            input_name: dataframe[input_name].values
            for input_name in self.input_names
        }` fails. (The dataframe does not include a 'dense_1_input_01' column).
 For this reason, I made it a requirement that there should be only one input as `predict` accepts one dataframe as input. Any thoughts?

Reply:
@avflor This is a good point. Based on the suggestion you provided in your other comment, I think it makes sense to implement the following behavior for pyfunc inference:

1. In the case where the underlying ONNX model only has a single input, pass `df.values` to the ONNX model (as in the current state of the PR).

2. In the case where the underlying ONNX model has *multiple* inputs, attempt to select dataframe columns with the same names as the ONNX model inputs (as in my proposed solution above).

This proposal has the benefit of supporting multi-input ONNX models while maintaining ease of use for the most common case where the model has a single input.

I also considered the following alternatives:

1. **Keep the current state of the `_OnnxModelWrapper`, restricting the `pyfunc` representation to ONNX models with only one input**. In this case, it does not make sense to add the `pyfunc` flavor to ONNX models with multiple inputs because the `pyfunc` representation cannot be used to perform inference. However, it also seems confusing to omit the `pyfunc` flavor in these cases because MLflow will produce different sets of flavors for a model based on the number of ONNX inputs. This approach also has the disadvantage of restricting interoperability for a broad class of multi-input models.

2. **Use the tensorflow-style, column-based approach for evaluating all ONNX models as python functions, regardless of the number of ONNX inputs**. This API makes it challenging for users to feed multi-dimensional input values to models (e.g., an n-by-4 matrix of Iris data); it requires the user to embed a multi-dimensional array inside a dataframe column. While the current proposal still requires this nested list behavior for ONNX models with *multiple* inputs, the proposed `df.values` API for ONNX models with one input provides a much simpler API for specifying multidimensional data. Additionally, the column-based approach has the disadvantage of always requiring the user to find the names of all ONNX inputs for a model, even if the model only contains a single input. As you noted in the case of converted Keras models, ONNX input names are often autogenerated by various conversion APIs and are not particularly intuitive.

Please let me know if you have any questions or pushback about the proposed solution and the alternatives considered. If we agree to adopt this approach, it's important to thoroughly document the evaluation behavior and provide examples.

Reply:
@dbczumar I'm also in favor of the approach you proposed as it allows supporting a broader range of models without making the API more complex. I added both single-input and multi-input implementations during inference. Please take a look and let me know if you have any other comments.